### PR TITLE
Added neighbour position iterator.

### PR DESCRIPTION
**Pull request by Caellian**
NeighbourPositionIterator is a tiny utility class which iterates over block positions surrounding given center block position.

This would help me with network management in Flow API and would likely be useful to other mod developers.

Better alternative to this would be to have a big EnumFacing class which includes all (even diagonal) directions and has multiple arrays for iteration with those directions, but adding a new class would be pointless because vanilla Minecraft has it's own (which people should use) and patching Minecraft one is just calling for trouble because it could break any time and even when fixed (patching dozens of vanilla classes) new versions would likely break it again sooner than later. So sadly, this PR is the most logical choice considering the conundrum above.

Signed-off-by: Caellian <tin.svagelj.email@gmail.com>

### CODE REVIEW

File: src/main/java/xyz/openmodloader/util/NeighbourPositionIterator.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private static Vec3i[] allDirections = null;
+
+    private final BlockPos centerPosition;
+    private final boolean diagonalMode;
+    private int iterationStep = 0;
+
+    public NeighbourPositionIterator(BlockPos centerPosition, boolean includeDiagonalPositions) {
+        this.centerPosition = centerPosition;
+        this.diagonalMode = includeDiagonalPositions;
+        if (diagonalMode && allDirections == null){
```
Comment (Victorious3) [inline src/main/java/xyz/openmodloader/util/NeighbourPositionIterator.java]: What about moving this into a `static {}` block? Clear intention and less error prone.

**Comment by Victorious3** [comment]
I would add a small patch that adds `neighbors()` to `BlockPos` itself.

**Comment by Caellian** [comment]
@Victorious3 you wishes have been granted.
Any other suggestions? If no, merge please so I can use this class in my Power API branch.

File: patches/net/minecraft/util/math/BlockPos.java.patch [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ 
++    // OML
++    /**
++     * Creates an array of neighbour BlockPos objects.
++     *
++     * @param diagonal true if diagonal block positions should be added too.
++     * @return Neighbour BlockPos objects.
++     */
++    public BlockPos[] neighbours(boolean diagonal){
++        ArrayList<BlockPos> result = Lists.newArrayList();
```
Comment (Victorious3) [inline patches/net/minecraft/util/math/BlockPos.java.patch]: If you are using Lists, you might as well use `Lists.newArrayList(iterator)`, bit less code...

**Comment by Caellian** [comment]
Wasn't aware of `Lists.newArrayList(iterator)`. Done! It's much smaller now.

**Comment by Caellian** [comment]
Should I maybe add getNeighbourBlocks(BlockPos pos) to World class?

File: patches/net/minecraft/util/math/BlockPos.java.patch [hunk 3]
```diff
@@ -0,0 +1,29 @@
+--- /net/minecraft/util/math
++++ /net/minecraft/util/math
+@@ -14,6 +14,7 @@
+ import org.apache.logging.log4j.Logger;
+ import xyz.openmodloader.launcher.strippable.Side;
+ import xyz.openmodloader.launcher.strippable.Strippable;
++import xyz.openmodloader.util.NeighbourPositionIterator;
```
Comment (lclc98) [inline patches/net/minecraft/util/math/BlockPos.java.patch]: We are trying not to have imports in patches, try putting this import below, like
 `List<BlockPos> result = Lists.newArrayList(new xyz.openmodloader.util.NeighbourPositionIterator(this, diagonal));`

**Comment by Caellian** [comment]
I already saw some imports so I assumed it's okay. I can fix it when I come home.

**Comment by Caellian** [comment]
@shadowfacts review and merge this if possible please, I need it to continue working on Power API.

**Pull request merged**